Backpack/Icon/Classes/BPKIcon.m

```diff
@@ -0,0 +1,236 @@
+
+    CGSize iconSize = [self concreteSizeForIconSize:size];
+
+    NSString *cacheKey = [self cacheKeyForIconNamed:name withColor:color size:size];
+    UIImage *icon = [self.imageCache objectForKey:cacheKey];
+
+    if (icon) {
+        return icon;
+    }
+
+    UIGraphicsBeginImageContextWithOptions(iconSize, NO, 0.0);
+
+    if (self.iconMapping) {
+        NSAssert([self.iconMapping objectForKey:name], @"Unknown icon `%@` does not correspond to a known icon", name);
+    }
+    NSString *iconCodepoint = [self stringForUnicodeCodepoint:[self.iconMapping objectForKey:name]];
+
+    UIFont *font = [UIFont fontWithName:@"BpkIcon" size:iconSize.height];
+    NSAssert(font, @"`BpkIcon` was not correctly registered.");
+
+    if (font == nil) {
+        font = [UIFont systemFontOfSize:iconSize.height];
+        iconCodepoint = BPKIconFallbackGlyph;
+    }
+
+    [iconCodepoint drawInRect:CGRectMake(0, 0, iconSize.width, iconSize.height)
+               withAttributes:@{
+                                NSFontAttributeName : font,
+                                NSForegroundColorAttributeName : color,
+                                NSParagraphStyleAttributeName:self.paragraphStyle,
+                                }];
+    icon = UIGraphicsGetImageFromCurrentImageContext();
+    UIGraphicsEndImageContext();
+
+    [self.imageCache setObject:icon forKey:cacheKey];
+
+    return icon;
+}
+
+#pragma mark - Private
+
++ (NSCache<NSString *, UIImage *> *)imageCache {
+    static NSCache<NSString *, UIImage *> *_imageCache;
+    static dispatch_once_t dispatchOnceCacheToken;
+
+    dispatch_once(&dispatchOnceCacheToken, ^{
+        _imageCache = [[NSCache alloc] init];
+    });
+
+    return _imageCache;
+}
+
++ (NSParagraphStyle *)paragraphStyle {
+    static NSParagraphStyle *_paragraphStyle;
+    static dispatch_once_t dispatchOnceParagraphStyleToken;
+
+    dispatch_once(&dispatchOnceParagraphStyleToken, ^{
+        NSMutableParagraphStyle *mutableStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];
+        mutableStyle.alignment = NSTextAlignmentCenter;
+
+        _paragraphStyle = [mutableStyle copy];
+    });
+
+    return _paragraphStyle;
+}
+
++ (nullable NSDictionary<NSString *, NSString *> *)iconMapping {
+    static NSDictionary<NSString *, NSString *> *_iconMapping;
+    static dispatch_once_t dispatchOnceiconMappingToken;
+
+    dispatch_once(&dispatchOnceiconMappingToken, ^{
+        NSBundle *bundle = [self iconBundle];
+        if (bundle == nil) {
+            _iconMapping = nil;
+            return;
+        }
+
+
+        NSURL *url = [bundle URLForResource:@"iconMapping" withExtension:@"json"];
+        NSAssert(url, @"`iconMapping.json` must be present in the bundle.");
+        if (url == nil) {
+            _iconMapping = nil;
+            return;
+        }
+
+        NSData *data = [NSData dataWithContentsOfURL:url];
```

Comment:
Could have used `dataWithContentsOfURL:options:error:` for more accurate error reporting

Reply:
Sounds good yeah